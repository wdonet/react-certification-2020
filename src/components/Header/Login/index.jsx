import React from 'react';
import wizelogo from '../../../img/wize-logo.png';

const Login = () => {
  const authenticated = true;

  return (
    <a className="login-link" href="_blank">
      <img src={wizelogo} alt="wizelogo" height={60} />
      {authenticated ? <div className="login-user">Mario</div> : <div>Sign in</div>}
    </a>
  );
};

export default Login;
