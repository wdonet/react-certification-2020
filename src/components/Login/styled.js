import styled from 'styled-components';

const LoginBackground = styled.div`
  display: none;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0, 0, 0);
  background-color: rgba(0, 0, 0, 0.7);

  display: ${({ open }) => (open ? 'block' : 'none')};
`;

const LoginBox = styled.div`
  background-color: #fefefe;
  margin: 15% auto;
  padding: 20px;
  border: 1px solid #888;
  border-radius: 5%;
  width: 300px;
  height: 350px;
  position: relative;
`;

const CloseLoginButton = styled.span`
  color: #aaa;
  position: absolute;
  top: 0px;
  right: 15px;
  font-size: 28px;
  font-weight: bold;
  &:hover,
  &:focus {
    color: black;
    text-decoration: none;
    cursor: pointer;
  }
`;

const ErrorText = styled.p`
  color: red;
  font-size: 14px;
`;

const InputText = styled.input`
  width: 100%;
  height: 30px;
  border: 1px solid ${({ theme }) => theme.font.secondary};
  margin: 10px 0;
`;

const FormStyled = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 20px 0;
`;

export { LoginBackground, LoginBox, CloseLoginButton, ErrorText, InputText, FormStyled };
