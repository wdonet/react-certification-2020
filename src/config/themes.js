export default {
  dark: {
    icon: 'moon',
    sound: 'https://www.joshwcomeau.com/sounds/switch-on.mp3',
    background: {
      primary: '#161f23',
      secondary: '#1e2931',
    },
    decorator: {
      primary: '#2c0e7b',
      secondary: '#194c7c',
      contrast: '#FFF',
    },
    text: {
      primary: '#FFF',
      secondary: '#FFF',
    },
  },
  light: {
    icon: 'sun',
    sound: 'https://www.joshwcomeau.com/sounds/switch-off.mp3',
    background: {
      primary: '#eef4f4',
      secondary: '#d7dee7',
    },
    decorator: {
      primary: '#5b8cbd',
      secondary: '#d29b3b',
      contrast: '#222',
    },
    text: {
      primary: '#222',
      secondary: '#222',
    },
  },
}
