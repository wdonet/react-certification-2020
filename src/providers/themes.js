const defaultColors = {
  white: '#FFFFFF',
  black: '#000000',
};

const font = {
  h1: {
    fontSize: '',
    fontWeight: '',
    letterSpacing: 1.5,
  },
};

export const lightTheme = {
  font,
  color: {
    ...defaultColors,

    primary: '#dce775',
    primaryVariant: '#c0ca33',
    secondary: '#c8ff00',

    background: '#ffffff',
    surface: '#f5f5f5',
    error: '#e64a19',
  },
};

export const darkTheme = {
  font,
  color: {
    ...defaultColors,

    primary: '#00695c',
    primaryVariant: '#004d40',
    secondary: '#1de9b6',

    background: '#616161',
    surface: '#757575',
    error: '#e65100',
  },
};
