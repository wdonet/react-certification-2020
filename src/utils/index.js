export { default as contextWrapper } from './testing/helpers/contextWrapper';
export { default as renderWithTheme } from './testing/helpers/renderWithTheme';
export { default as googleMockedAPIObject } from './testing/mocks/googleMockedAPIObject';
export { default as youtubeMockedData } from './testing/mocks/youtubeMockedData';
export { default as YTMockedObject } from './testing/mocks/YTMockedObject';
