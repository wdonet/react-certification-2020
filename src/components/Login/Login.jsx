import React, { useContext, useState } from "react";
import styled from "styled-components";
import loginService from '../../services/loginService';
import AppContext from "../../providers/AppContext";
import TextField from "../../ui/TextField/TextField";
import Button from "../../ui/Button/Button";
import { useHistory } from "react-router";

const StyledLogin = styled.div`
    padding: 4px;
    margin: 4px;
    height: min-content;
    width: 400px;
    background: ${ ({theme}) => theme.color.surface };
    color: ${ ({theme}) => theme.color.fontPrimary };
`

const StyledSection = styled.div`
    display: ${({display}) => display ? display : 'grid'};
    height: min-content;
    margin: 8px;
`;

const Login = (props) => {
    const { push } = useHistory();
    const { theme, setUserSession } = useContext(AppContext);
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const { onCancel } = props;

    const loginAttempt = async () => {
        const response = await loginService(username, password);
        setUserSession(response);
        push({ pathname: "/home" });
    }

    return (<StyledLogin role="form" theme={theme} data-testid="login-form">
                <StyledSection>Inicia sesión!</StyledSection>
                <StyledSection>
                    <label htmlFor="username">Usuario</label>
                    <TextField 
                        id="username" 
                        name="username"
                        data-testid="username-input"
                        onChange={(uname) => setUsername(uname)}
                    />
                </StyledSection>
                <StyledSection>
                    <label htmlFor="password">Contraseña</label>
                    <TextField 
                        id="password" 
                        name="password"
                        data-testid="password-input"
                        type="password"
                        onChange={(pword) => setPassword(pword)}
                    />
                </StyledSection>
                <StyledSection display="flex">
                    <Button 
                        data-testid="login-button"
                        onClick={loginAttempt}
                    >
                        Ingresar
                    </Button>
                    <Button 
                        data-testid="cancel-button"
                        onClick={() => onCancel && onCancel()}
                    >
                        Cancelar
                    </Button>
                </StyledSection>
            </StyledLogin>);
}

export default Login;