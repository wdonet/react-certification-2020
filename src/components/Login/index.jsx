import React, { useEffect, useRef, useState } from 'react';

import {
  LoginBackground,
  LoginBox,
  CloseLoginButton,
  ErrorText,
  InputText,
  FormStyled,
} from './styled';

import { useAuth } from '../../providers/Auth';

const Login = ({ open, closeModal }) => {
  const [loginError, setLoginError] = useState(null);
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const { login } = useAuth();
  const loginBoxRef = useRef(null);
  useEffect(() => {
    const handleClickOutside = (e) => {
      if (loginBoxRef.current && !loginBoxRef.current.contains(e.target)) {
        closeModal();
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [loginBoxRef]);

  const loginUser = async () => {
    const { success } = await login({ username, password });
    if (!success) {
      setLoginError('Invalid username or password');
    } else {
      setLoginError(null);
      closeModal();
    }
  };

  const handleUsernameChange = (e) => {
    const {
      target: { value },
    } = e;
    setUsername(value);
  };

  const hanglePasswordChange = (e) => {
    const {
      target: { value },
    } = e;
    setPassword(value);
  };

  return (
    <LoginBackground open={open}>
      <LoginBox ref={loginBoxRef}>
        <CloseLoginButton onClick={closeModal}>&times;</CloseLoginButton>
        <FormStyled>
          <InputText
            name="username"
            value={username}
            placeholder="Username"
            onChange={handleUsernameChange}
          />
          <InputText
            name="password"
            placeholder="Password"
            type="password"
            value={password}
            onChange={hanglePasswordChange}
          />
          {loginError && (
            <ErrorText
              css={`
                color: ${({ theme }) => theme.palette.alert.main};
                padding: 2px;
                position: absolute;
                top: calc(100% - 25px);
                font-size: 16px;
              `}
            >
              {loginError}
            </ErrorText>
          )}
          <button
            type="button"
            onClick={() => {
              loginUser({ username: 'fulano', password: '1234567' });
              return false;
            }}
          >
            Login
          </button>
        </FormStyled>
      </LoginBox>
    </LoginBackground>
  );
};

export default Login;
