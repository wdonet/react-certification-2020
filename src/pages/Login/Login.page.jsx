import React, { useState } from 'react';

import { useAuth } from '../../providers/Auth';
import { Container, ErrorLabel } from './Login.page.styled';

function LoginPage() {
  const { login } = useAuth();
  const [values, setValues] = useState({});
  const [error, setError] = useState();

  const authenticate = async (event) => {
    event.preventDefault();
    try {
      setError(undefined);
      await login(values);
    } catch (err) {
      setError(err);
    }
  };

  const handleInputChange = (event) => {
    const newValues = { ...values };
    newValues[event.target.id] = event.target.value;
    setValues(newValues);
  };

  return (
    <Container>
      <h1>Welcome back!</h1>
      <form onSubmit={authenticate} className="login-form">
        <div className="form-group">
          <label htmlFor="username">
            <strong>username </strong>
            <input
              required
              type="text"
              id="username"
              value={values.username || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <div className="form-group">
          <label htmlFor="password">
            <strong>password </strong>
            <input
              required
              type="password"
              id="password"
              value={values.password || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <button type="submit">login</button>
      </form>
      {error && <ErrorLabel>Username or password incorrect!</ErrorLabel>}
    </Container>
  );
}

export default LoginPage;
