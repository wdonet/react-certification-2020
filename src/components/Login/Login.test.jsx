import React from 'react';
import { act, render } from '@testing-library/react';
import Login from './Login'
import { contextWrapper, routerWrapper } from '../../utils';
import { lightTheme } from '../../providers/themes';
import AppContext from '../../providers/AppContext';

const build = async (Component = <Login/>) => {
    const contextValue = { theme: lightTheme, setUserSession: jest.fn() };
    let container;
    await act(async () => {
        const wrappedContext = contextWrapper(AppContext, contextValue, Component);
        const routeWrap = await routerWrapper(wrappedContext);
        container = render(routeWrap.wrap).container;
    });
    return { 
        container,
    };
};

describe('Login screen', () => {
    it('renders with default properties', async () => {
        const built = await build();
        const { firstChild } = built.container;
        expect(firstChild).toMatchSnapshot();
        expect(firstChild).toHaveStyle('border-radius: 8px');
        expect(firstChild).toHaveStyle('width: 400px');
        expect(firstChild).toHaveStyle('height: min-content');
        expect(firstChild).toHaveStyle('margin: 4px');
        expect(firstChild).toHaveStyle('padding: 4px');
    })
})