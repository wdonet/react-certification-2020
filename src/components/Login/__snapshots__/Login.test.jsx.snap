// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen renders with default properties 1`] = `
<div
  class="sc-htpNat eiVhmM"
  data-testid="login-form"
  role="form"
>
  <div
    class="sc-bxivhb dzcmIQ"
  >
    Inicia sesión!
  </div>
  <div
    class="sc-bxivhb dzcmIQ"
  >
    <label
      for="username"
    >
      Usuario
    </label>
    <input
      class="sc-bdVaJa fkGYjX"
      data-testid="username-input"
      id="username"
      name="username"
    />
  </div>
  <div
    class="sc-bxivhb dzcmIQ"
  >
    <label
      for="password"
    >
      Contraseña
    </label>
    <input
      class="sc-bdVaJa fkGYjX"
      data-testid="password-input"
      id="password"
      name="password"
      type="password"
    />
  </div>
  <div
    class="sc-bxivhb ccxHxs"
    display="flex"
  >
    <button
      class="sc-bwzfXH mRSDS"
      data-testid="login-button"
    >
      Ingresar
    </button>
    <button
      class="sc-bwzfXH mRSDS"
      data-testid="cancel-button"
    >
      Cancelar
    </button>
  </div>
</div>
`;
